buster/chatbot.py

```diff
@@ -121,59 +121,64 @@ def rank_documents(
 
         return matched_documents
 
-    def prepare_prompt(self, question: str, candidates: pd.DataFrame) -> str:
+    def prepare_documents(self, matched_documents: pd.DataFrame, max_words: int) -> str:
+        # gather the documents in one large plaintext variable
+        documents_list = matched_documents.text.to_list()
+        documents_str = " ".join(documents_list)
+
+        # truncate the documents to fit
+        # TODO: increase to actual token count
+        word_count = len(documents_str.split(" "))
+        if word_count > max_words:
+            logger.info("truncating documents to fit...")
+            documents_str = " ".join(documents_str.split(" ")[0:max_words])
+            logger.info(f"Documents after truncation: {documents_str}")
+
+        return documents_str
+
+    def prepare_prompt(
+        self,
+        question: str,
+        matched_documents: pd.DataFrame,
+        text_before_prompt: str,
+        text_before_documents: str,
+    ) -> str:
         """
         Prepare the prompt with prompt engineering.
         """
+        documents_str: str = self.prepare_documents(matched_documents, max_words=self.cfg.max_words)
+        return text_before_documents + documents_str + text_before_prompt + question
 
-        max_chars = self.cfg.max_chars
-        text_before_prompt = self.cfg.text_before_prompt
-
-        documents_list = candidates.text.to_list()
-        documents_str = " ".join(documents_list)
-        if len(documents_str) > max_chars:
-            logger.info("truncating documents to fit...")
-            documents_str = documents_str[0:max_chars]
+    def get_gpt_response(self, **completion_kwargs):
+        # Call the API to generate a response
+        logger.info(f"querying GPT...")
+        try:
+            return openai.Completion.create(**completion_kwargs)
 
-        return documents_str + text_before_prompt + question
+        except Exception as e:
+            # log the error and return a generic response instead.
+            logger.exception("Error connecting to OpenAI API. See traceback:")
+            response = {"choices": [{"text": "We're having trouble connecting to OpenAI right now... Try again soon!"}]}
+            return response
 
     def generate_response(self, prompt: str, matched_documents: pd.DataFrame) -> str:
         """
         Generate a response based on the retrieved documents.
         """
         if len(matched_documents) == 0:
             # No matching documents were retrieved, return
-            response_text = "I did not find any relevant documentation related to your question."
-            return response_text
+            return "I did not find any relevant documentation related to your question."
```

Comment:
Seems redundant since it's also in format_response

Reply:
hmm youre right, it shouldnt be in both, let me fix that

Reply:
i just changed it to something more "clever". If the length of documents is 0, we will instead return the `unknown_prompt` text, which will then always trigger the UNK token in the `is_relevant` function, bypassing the need to check for the length.